### ISSUE TITLE
No Profile Mode - issues and questions.

### ISSUE THREAD

**Issue by nishihatapalmer** [digital-preservation/droid#375]
The No Profile mode in the command line project suffers from multiple issues.

- doesn't match extensions the same as profile mode.
- only outputs filename and puid.
- archive file processing support is limited (it usually can't handle nested archives)

**Origin**
The original idea of no profile mode was to support outputting droid metadata without the overhead of a database.  If the results were just going to go to the standard out, or a file, then there was no need to record them in a database as well.

**Implementation**
A decision was made to re-implement all the core identification and archive handling in the command line project for the no profile mode.  This was a short term decision made due to some annoying incompatibilities (filters being one).  At the time there was only zip, gz and tar archive handling.

**Outcomes**
The result of this is that there is a considerable maintenance burden imposed by the no profile mode, as the archive handling must be re-implemented every time a new type is added.  In fact, the implementations provided by no profile mode are much less capable than the ones already available in the main droid projects (they can't handle nested archives for the most part).

Identification also suffers, as the logic doesn't match what is available in the rest of DROID, and results suffer as very little metadata is outputted in this mode (just filename and puid).

**Original design**
In the original design, the idea of no profile mode was that it should use all the same profile machinery, but just plug in a csv output ResultsHandler (whether to standard out or a file), instead of the database results handler.

**Questions**
1.  Would implementing no profile mode to the original design be of interest?  Doing so would massively simplify the command line project, as it would just use the standard droid modules used by the rest of DROID.  
2.  Do people depend on the current output (just filename and puid) of no profile mode?  This could be preserved, although I suspect most people will want the full metadata most of the time.

_Note_: I see there is an issue with no-profile-mode filenames containing commas here: https://github.com/digital-preservation/droid/issues/34  This would be fixed by using a CSV output format for droid metadata.  The idea would be that no profile mode could output data essentially the same as if we had profiled with a database, and then performed an export.  The output could still be given in the current format for backwards compatibility reasons.

**Comment by nishihatapalmer** [comment]
There is one unavoidable difference between outputting results in real-time and using a database.  If a previous result has to be updated, then the database will be fine, but it's not clear what to do with continual output.  

This can happen when an archival format is identified, but then an error occurs processing the parts of it.  The original identification won't have the error status.  Error reporting in continual output is something that needs a bit more consideration.

**Comment by nishihatapalmer** [comment]
The only approaches I can see to error reporting are:

1.  Defer recording the identification metadata of the parent archival format until we've successfully processed it.  This would lead to archival parents always being recorded after their children in the output.
2.  Allow error reporting entries in the output.  These are corrections which should be applied to the status of previous entries.

**Comment by marhop** [comment]
> Would implementing no profile mode to the original design be of interest?

Yes! It would be great to have consistent results with CLI (no profile) and GUI DROID. See also #224 which I think is related.

**Comment by thorsted** [comment]
I also would have interested in seeing consistency between the two versions. I would first want to know how this would affect our enterprise installation of Rosetta and how it implements DROID. I believe it uses the CLI version, but also has added a way to set the Max Byte Scan size setting. It would also allow us to add the ability to scan more archival formats within the Rosetta preservation system.

**Comment by jcharlet** [comment]
@nishihatapalmer we just have been thinking that we should really refactor the whole archive management. 
I had to fix a few inconsistencies between GUI and CLI like in https://github.com/digital-preservation/droid/pull/378 and https://github.com/digital-preservation/droid/pull/376, and this is truly a maintenance burden.

**Comment by nishihatapalmer** [comment]
I'm having a play with what's needed for this.  So far looks mostly like implementing a WriterResultHandlerDao class in the results.handlers package.

**Comment by jcharlet** [comment]
really looking forward to it Matt. I think it should be prioritized for next release, if you can do it, that would be amazing otherwise we'll work on it with @sparkhi .

**Comment by nishihatapalmer** [comment]
Well, definitely not this release!  

Turns out profiles have a few other database dependencies (like reporting and exporting), although they all live behind interfaces, and they aren't needed for direct output to console or file.  More to explore.

**Comment by jcharlet** [comment]
yeah, this is definitely not trivial.

**Comment by nishihatapalmer** [comment]
**Update**

I've successfully run DROID exporting CSV results to the console, by just implementing a CSV result handler dao class and using it instead of the JDBC one.  So the principle seems solid.

There's some additional refactoring requiring to obtain the list of formats (the database result handler can do this, the CSV one can't, so we need to change how formats are acquired to make this generic).

**Comment by nishihatapalmer** [comment]
Turns out refactoring how formats are acquired is a hugely complex.  It hits into all kinds of horrid database initialisation problems and init fixup methods apparently caused by problems with spring not initialisating things in the order you'd hope.

 Instead of refactoring half the code base to acquire them sanely, it's going to be easier to just add a database dependency for format loading to the CSV result handler dao (we can abstract the common functionality out into an abstract base class).  It's not how I would design it given a clean slate, but it's the least disturbing option.

**Comment by nishihatapalmer** [comment]
@adamretter @jcharlet I have a spring question for you guys, as I'm really not an expert in it.

I have two result handler dao beans - one for JDBC output to database, and one to output CSV to standard out or a file.  What's a good spring way to select the right bean to be used during profile initialization?

Right now, the JDBCResultHandlerDao bean is set directly in the spring-results.xml file.  Should I create a factory class with properties passed in that select which bean is ultimately used?  Or is there some other canonical spring way to select which bean is ultimately used?

**Comment by nishihatapalmer** [comment]
Current code is in my fork at https://github.com/nishihatapalmer/droid/tree/NoDatabaseProfiling

**Comment by jcharlet** [comment]
@nishihatapalmer I prefer dependency injection from annotation, but it's the same as the one in place using xml. I wouldn't do anything smarter than you do already, probably inject the two beans in a service using the interface while giving the implementation (bean) name in the injection annotation, then use one or the other based on some condition. factory class sounds better, if it's not creating too much boilerplate code, in my humble opinion.

**Comment by jcharlet** [comment]
right, the initialisations in the handler are... complex. 
So in the case of the GUI, we would use both handlers right, the jdbc one to enable all db related features (export, reporting), and the writerResultsHandlerDao for displaying the results?

**Comment by nishihatapalmer** [comment]
Yes, the whole dB initialisation is pretty hairy.

The idea would be that the JDBC would be used in the GUI, and the Writer handler would be used in the command line to enable pushing results directly to std out or a file, rather than a database.

Not much need for direct file output in the GUI I suspect (although it is a possibility).

**Comment by adamretter** [comment]
@nishihatapalmer I am afraid that I stopped working with Spring before I even contributed to DROID, so I am probably not the best person to ask about "Spring Best Practice" ;-)

**Comment by nishihatapalmer** [comment]
@adamretter A wise decision IMHO!  I am in general not a fan of huge frameworks.  They make easy things trivial, medium things hard, and hard things impossible!  You spend more time fighting the framework and twisting the code to fit it than you gain by using it.

I'm sure there are useful frameworks out there.  I have just found all the big all encompassing ones to be more trouble than they're worth.

**Comment by nishihatapalmer** [comment]
Managed to wire things in using a FactoryBean and a new profile property {$outputFilePath}.

If "stdout", it writes to the console, otherwise it creates a file at the file path provided.

Next step is to try this in the command line project.

**Comment by jcharlet** [comment]
Hi @nishihatapalmer , we will start our new work on DROID by auditing the functionality and code discrepancies between GUI and CLI, to then merging the common codebase between them and fixing those inconsistencies.
Any code or thoughts you'd like to share on this on top of those in that thread?

**Comment by nishihatapalmer** [comment]
Only a few comments; I haven't been coding much for the last few months, what with everything that's going on right now.  I'll have to dig out all my old code to give a proper update.  

To summarise where I got to:

1. I still believe the correct way to merge the code bases is to essentially ditch most of the CLI code and make it use the code used by the GUI project.  This was always the original intention.

2. I have created code which replaces the database writer with a CSV file writer, which works in the GUI by specifying a new profile property $outputFilePath.  This is really easy to do and involves almost no new code.

3. I have realised that so many things assume that a profile has a database attached to it, that it's not worth trying to remove databases from profiles.  It's simply easier to always have an empty database with a profile,  but just to write out the results to a file instead of the database.

4. The reason I submitted a the PR https://github.com/digital-preservation/droid/pull/407 to allow the command line to set profile properties was because in order for the command line to use the profile code, it needs to set the $outputFilePath property.  This also has the nice effect of allowing the command line to override things like maxBytesToScan and things like that.

Next steps

5.  On its own, this still wasn't enough, as $outputFilePath isn't currently a global profile property - there are additional properties passed to the database on top of the global properties, but the code path by which this happens is rather tied to the GUI and database initialisation.  

6. I've played with expanding the command line properties to allow any property to be passed in just like the GUI, but this runs into all sorts of horrid bits of code and assumptions.  Maybe when I look at it again, it will all click into place and it will be easy... An alternative strategy is just to make the one property I *really* need to pass ($outputFilePath) a global property and be done with it.   This is less flexible, but much, much easier to implement.

Summary
I'm close to being able to use all the profile machinery from the command line writing out to a file instead of to the database.  At that point, it would be possible to ditch the command line implementations of container matching and archive handling.

Backwards Compatibility
The current no profile mode outputs just the path and PUID.  Do we need to preserve that behaviour, or would we just have a CSV output?  I suspect most people will actually want more of the metadata, but we may need to consider a backwards compatible output mode.  This would be very easy to do.

**Comment by nishihatapalmer** [comment]
I'll dig through my code to see what branches I have that I haven't pushed to my fork.  Then you'll be able to review what I have.  None of it is quite ready to be a PR yet.

**Comment by jcharlet** [comment]
to be fixed with https://github.com/digital-preservation/droid/issues/470

**Comment by jcharlet** [comment]
Brilliant Matt. We are now focusing on refactoring the CLI and GUI code, so this will be really helpful and we can carry on from there.

**Comment by jcharlet** [comment]
Hi @nishihatapalmer 
Been trying to find where to override the bean definition of resultsDao to put the writer instead of the db one when using the CLI.

There are spring contexts for each module like ui-spring.xml which I expected they would import spring contexts of submodules like spring-results.xml. But instead, they register factory beans (like SpringProfileInstanceFactory) which when called, generate the modules beans contexts. Why make this so? Isn't it causing the issues you're having in #407 with some properties not being easily overridden? 
I understand you want to instantiate some beans for every new profile which are maybe not thread safe, but what about daos, shouldn't they be singletons registered once and for all in the app context?

I couldn't find your $outputFilePath property you mentioned being used in the GUI, where can I find that? But also I'm understanding for your latest comment on #407 that you haven't sorted this out just yet am I right? 
So in the current state, #407 won't help just yet with this issue. Do you have any idea how to use this resultsDao otherwise? I'll carry on searching this afternoon.


Matt you complained about Spring framework above, although I don't have your much larger experience, I agree with the fact that they make uncommon things hard to implement, but I strongly disagree on the overall benefits. If the project was running on springboot, we wouldn't be having this conversation, properties overriding through command line would be available out of the box. Context management would be sorted with a couple of conditional annotations on beans. If we don't abandon entirely some responsibilities to the framework (like IoC), and do a mix of custom code and framework use like here, we're bound to get in trouble, lose the benefits of using the framework, and make our lives harder indeed for anything involving the framework.
I also think than keeping some of those libraries with a higher level of abstraction in place (rather than rewriting stuff at a lower level) make the application easier to maintain over time and might still be favourable to small performance gains, which feels paramount to me in this context of a project with both a very long life expectancy and a high developer turn over. 

Not sure if this is the right place to debate on this, I should probably have bounced back right away on your conversation with Adam above. Matt, @adamretter , I'm really keen to hear your thoughts on this and better understand from your perspectives.

**Comment by nishihatapalmer** [comment]
Hi @jcharlet ,

I didn't write the spring implementation used in DROID and I'm certainly no spring expert.  In fact I tend to dislike huge frameworks in general as I always seem to end up fighting them.  Maybe I'd learn to love springboot, who knows?

The outputFilePath property was introduced in a couple of later commits (I've now pushed them to my branch), along with a FactoryBean to instantiate the right result Dao.  You could only use it by hard coding this property, it wasn't properly wired in yet (in fact, the GUI is a red herring, it's just what I was using to test it).  But if you set this property, you can make the profile output to a file rather than the database.  There is a reason why this branch isn't yet a PR!

I've now pushed a few more commits that allow a ProfileInstance to set the outputFilePath property. My next step is to combine this with being able to set command line profile properties.  If we can then set the outputFilePath via a command line property, I can begin the work of removing the special archive and container signature processing in the command line project.

**Comment by nishihatapalmer** [comment]
I've also merged in the latest master changes and pushed that all up too, it was quite behind.

**Comment by nishihatapalmer** [comment]
If you ever wanted to refactor the spring implementation in DROID to be sane, I'd love to see that.  It does seem to manage to make simple things hard and hard things impossible, which I'm thinking isn't generally the idea of using things like spring.

My preferred method would just be to get rid of it entirely and instantiate everything in code with constructor injection, but then I've made my feelings on big frameworks known already!

**Comment by jcharlet** [comment]
hi @nishihatapalmer ,

I'm sorry, I think Richard set up that spring system, and I'm sure he had really good reasons to follow that pattern in the first place. I got frustrated because it's nothing like I saw, and that makes this simple task way more complicated than it should.

haha indeed you made it very clear :), and I really agree with you, in the current state, Spring is not helping, on the contrary, I really don't like this setup. I am not even sure it would help to setup springboot since many of the things that it abstracts away are already implemented. But it's a good point, there is space for refactoring, maybe removing the Spring context initialisation might simplify things, but if I do it, I have to see how we can include that in our priorities.

Ok great thanks for those additions. Would it help if I merge the properties PR, and merge it into your branch? I know it's not enabling setting properties used in droid-results, but it can maybe get us a step closer.

**Comment by jcharlet** [comment]
Ah nevermind I saw your comments on the PR I created, let's carry on the discussion there

**Comment by nishihatapalmer** [comment]
Yes, they do need to be merged at some point.  I'm not a git expert though, so I will be guided by you.

My plan was to create another branch off my "no database profiling" branch (just to be safe), and merge in the profile properties PR branch.  This would give me a merged branch, which I could then submit as a new PR once it was ready.  

But... you have already created a PR from my incomplete branch, so I'm not sure what the best strategy is.  I guess if all the same code is merged, it won't be hard to reconcile (plus or minus a few minor merge conflict differences I guess)...

**Comment by jcharlet** [comment]
I'd rather merge your properties PR first, which I'm working on (but I'm facing random unit test fails on the CI with appveyor).

then merge/rebase master (including that PR) on your branch. Sorry I'm a bit picky and I prefer to rebase master on feature branches (rather than merge, to keep git history graph clearer), and I didn't want to risk losing stuff in the merge conflicts' resolution, this is why I created my own branch. Then I cherry picked your new commits. but we can just as simply use your branch, merge master on it, and create a PR from it (I will close the one I did)

**Comment by nishihatapalmer** [comment]
OK, I'll wait for you to work out whatever problems you're having.  Let me know how I get all this merged in my local copy so I can continue to work on it.

**Comment by nishihatapalmer** [comment]
I'm basically confused by you making a PR out of an incomplete branch of mine.  Mind blown along with rebasing, etc. Let me know when it's safe for me to work on my code again and how I go about doing it!

**Comment by jcharlet** [comment]
I took over your work because we decided it was our priority to refactor the inconsistencies between the GUI and CLI. It was better to start from your incomplete branch, rather than start from scratch.

If you are available and want to work on this, I leave it to you. 
I don't understand what you don't understand about rebasing, but if it's confusing, fine, I'll remove my branch and my PR, and merge master on your branch now.

**Comment by jcharlet** [comment]
The point of making a "draft PR" is to have a place to discuss code before the work is complete.https://github.blog/2019-02-14-introducing-draft-pull-requests/

The point of rebasing vs merging
https://timwise.co.uk/2019/10/14/merge-vs-rebase/

**Comment by jcharlet** [comment]
@nishihatapalmer You can work back on your branch.
Master is merged on it.

I had to manage that conflict, and I added that todo in ProfileContextLocator:

```
        //TODO manage OUTPUT_FILE_PATH overriding
//        profileInstance.setOutputFilePath(mergedConfig.getString(DroidGlobalProperty.OUTPUT_FILE_PATH.getName()));
        profileInstance.setOutputFilePath(globalConfig.getProperties().getString(OUTPUT_FILE_PATH));
```

**Comment by nishihatapalmer** [comment]
Hi @jcharlet,

Thanks for accomodating me, it is appreciated.  

I seem to remember from several years ago is that rebase was supposed to be bad because it rewrites history, and makes it hard to collaborate.  I guess times have moved on and rewriting history is now a better option in some cases.  In any case, I'm not too bothered what the right approach is, I'll leave that to people who really care about such things - I just wanted to get on with doing some coding! 

Many thanks for your help and working around my limited understanding.

**Comment by jcharlet** [comment]
Well I guess it depends how many conflicts there are and how it makes the rebase/merge complicated :). I find it way easier to troubleshoot and come back in time when you have a clear history with mainly one continuous branch including all works, this is why I favour rebase whenever it's possible. 

And of course, I should be accommodating, I am way too grateful for your work on Droid! My mistake, I should have taken the time to explain better the first time I started using rebasing on your branches.

**Comment by nishihatapalmer** [comment]
Update - last year was a complete write off for me writing any new code.  New job, pandemic, other stuff.

I'm now back looking at this code again, and I've just run the first successful CSV output from the command line using all the standard profile code, instead of the command line versions for archive and container processing.

Needs testing, and I'll have to look at making a bit more configurable so we can duplicate the "No Profile" output as closely as possible (e.g. not all the metadata, don't quote the fields).  Then I can alias the old command to this new code, and we'll be able to remove the unnecessary duplication of code in the command line project.

**Comment by nishihatapalmer** [comment]
OK - I've now committed code to my fork that gets rid of the no profile code.  It also adds a number of new features into the profiling code (configuration of columns to export, quoting of columns to export, set arbitrary signature files) that were required to alias the old no profile code to be as close as possible to its original behaviour.

The command line code is now very clean and small.  A number of bugs in both command line and GUI have also been fixed along the way.

Needs more testing before I submit a PR, but should be fairly soon.

**Comment by nishihatapalmer** [comment]
Update, since it's all gone a bit quiet.  It's been busy at work, so I haven't had a lot of time until now.  
Still need to do more testing and write a few more tests.  I plan to add a few more unit tests to get the coverage of the command line project up a bit, and to also execute a manual test plan to validate that all commands work as expected, which I'll share.

**Comment by nishihatapalmer** [comment]
I hope to be PR ready around mid June, possibly before.

**Comment by nishihatapalmer** [comment]
I'm basically ready to submit a PR.  I just want to run a few more tests to investigate a couple of things.
However - I'm currently blocked by the recent changes to master breaking build.  I can build in maven now, but for some reason IntelliJ refuses to.  Once I get back to being able to build the code, I'll run the last few tests and submit the PR.
